File: LC0743_Network_Delay_Time/network_delay_time.py
import heapq
from typing import List

class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # Construct a directed graph, where the cost of an edge is its signal travel time.
        adj = {} # maps node # to a list of its outgoing edges, denote by (target node, travel time)
        visited = {i + 1: False for i in range(n)} # denotes if a node is visited
        min_times = {i + 1: float('inf') for i in range(n)} # tracks minimum time for each node to receive the signal sent by node k
        for edge in times:
            u, v, w = edge
            if u not in adj:
                adj[u] = []
            if v not in adj:
                adj[v] = []
            adj[u].append((v, w))
        #print("adj:", adj)
        # Perform Dijkstra's to find the minimum time for each node to receive a signal sent by node k.
        pq = []
        heapq.heappush(pq, (0, k))
        while pq:
            curr_time, curr_node = heapq.heappop(pq)
            if not visited[curr_node]:
                visited[curr_node] = True
                min_times[curr_node] = curr_time
                # Enqueue all the unvisited neighbors of curr_node.
                for out_edge in adj[curr_node]:
                    neighbor, cost = out_edge
                    if not visited[neighbor]:
                        heapq.heappush(pq, (min_times[curr_node] + cost, neighbor))
        #print("min_times:", min_times)
        # Scan through min_times to find the longest time it takes any one node to receive the signal.
        answer = float('-inf')
        for node, time in min_times.items():
            if time == float('inf'):
                return -1
            answer = max(answer, time)
        
        return answer
        

File: LC0743_Network_Delay_Time/network_delay_time.go
package main

import (
	"container/heap"
	"math"
)

const INF = math.MaxInt32

type Pair struct {
	Key   int // used for ordering (min heap)
	Value int
}

type MinHeap []Pair

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].Key < h[j].Key }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MinHeap) Push(x any) {
	*h = append(*h, x.(Pair))
}

func (h *MinHeap) Pop() any {
	old := *h
	n := len(old)
	item := old[n-1]
	*h = old[:n-1]
	return item
}

func networkDelayTime(times [][]int, n int, k int) int {
	adj := make(map[int][][2]int)
	visited := make(map[int]bool)
	min_times := make(map[int]int)
	for i := 1; i <= n; i++ {
		adj[i] = make([][2]int, 0)
		visited[i] = false
		min_times[i] = INF
	}

	// Construct graph, with cost of an edge as time it takes for a signal to travel that edge.
	for _, time := range times {
		u, v, w := time[0], time[1], time[2]
		adj[u] = append(adj[u], [2]int{v, w})
	}

	// Perform Dijkstra's from k
	pq := &MinHeap{}
	heap.Init(pq)

	heap.Push(pq, Pair{Key: 0, Value: k})

	for pq.Len() > 0 {
		x := heap.Pop(pq).(Pair)
		curr_time, curr_node := x.Key, x.Value

		if !visited[curr_node] {
			visited[curr_node] = true
			min_times[curr_node] = curr_time
			// Enqueue all the unvisited neighbors of curr_node.
			for _, edge := range adj[curr_node] {
				target, cost := edge[0], edge[1]
				if !visited[target] {
					heap.Push(pq, Pair{Key: curr_time + cost, Value: target})
				}
			}
		}
	}

	// Scan through min_times to find the longest times it takes any one node to receive the signal.
	answer := -1
	for _, time := range min_times {
		if time == INF {
			return -1
		}
		if time > answer {
			answer = time
		}
	}

	return answer
}
